src/Data/Singletons/Prelude/List.hs

```diff
@@ -306,12 +306,18 @@ $(singletonsOnly [d|
   prependToAll _   []     = []
   prependToAll sep (x:xs) = sep : x : prependToAll sep xs
 
-  permutations            :: [a] -> [[a]]
+  permutations            :: forall a. [a] -> [[a]]
   permutations xs0        =  xs0 : perms xs0 []
     where
       perms []     _  = []
       perms (t:ts) is = foldr interleave (perms ts (t:is)) (permutations is)
         where interleave    xs     r = let (_,zs) = interleave' id xs r in zs
+
+              -- This type signature isn't present in the reference
+              -- implementation of permutations in base. However, it is needed
+              -- here, since (at least in GHC 8.2.1) the singletonized version
+              -- will fail to typecheck without it. See #13549 for the full story.
+              interleave' :: ([a] -> b) -> [a] -> [b] -> ([a], [b])
```

Comment:
It turns out adding one type signature is enough to work around [GHC Trac #13549](https://ghc.haskell.org/trac/ghc/ticket/13549). Hooray!